services/powchain/src/main/java/tech/pegasys/teku/services/powchain/execution/ExecutionEngineService.java

```diff
@@ -57,45 +63,99 @@ public void chainHeadUpdated(
       Bytes32 currentDutyDependentRoot,
       Optional<ReorgContext> optionalReorgContext) {
 
-    recentChainData
-        .retrieveBlockByRoot(bestBlockRoot)
-        .thenCompose(
-            maybeABlock ->
-                maybeABlock
-                    .flatMap(block -> block.getBody().toVersionMerge())
-                    .map(this::onHeadBlockMerge)
-                    .orElseGet(() -> SafeFuture.completedFuture(null)))
-        .finish(err -> LOG.warn("setHead failed", err));
-  }
+    final Optional<Bytes32> curFinalizedBlock;
 
-  private SafeFuture<Void> onHeadBlockMerge(BeaconBlockBodyMerge blockBody) {
-    // Check if there is a payload
-    if (!blockBody.getExecution_payload().equals(new ExecutionPayload())) {
-      return executionEngineChannel.setHead(blockBody.getExecution_payload().getBlock_hash());
-    } else {
-      return SafeFuture.completedFuture(null);
+    synchronized (this) {
+      lastBestBlock = Optional.of(bestBlockRoot);
+      curFinalizedBlock = lastFinalizedBlock;
     }
+
+    updateForkChoice(bestBlockRoot, curFinalizedBlock);
   }
 
   @Override
   public void onNewFinalizedCheckpoint(Checkpoint checkpoint) {
-    recentChainData
-        .retrieveBlockByRoot(checkpoint.getRoot())
+    final Bytes32 curBestBlock;
+    final Optional<Bytes32> curFinalizedBlock;
+
+    synchronized (this) {
+      lastFinalizedBlock = Optional.of(checkpoint.getRoot());
+      if (lastBestBlock.isEmpty()) {
+        return;
+      }
+      curBestBlock = lastBestBlock.get();
+      curFinalizedBlock = lastFinalizedBlock;
+    }
+
+    updateForkChoice(curBestBlock, curFinalizedBlock);
+  }
+
+  private static Optional<Bytes32> getPowBlockHash(Optional<BeaconBlock> block) {
+    return block
+        .map(BeaconBlock::getBody)
+        .flatMap(BeaconBlockBody::toVersionMerge)
+        .map(BeaconBlockBodyMerge::getExecution_payload)
+        .filter(executionPayload -> !DEFAULT_EXECUTION_PAYLOAD.equals(executionPayload))
+        .map(ExecutionPayload::getBlock_hash);
+  }
+
+  private void updateForkChoice(Bytes32 bestBlockRoot, Optional<Bytes32> maybeFinalizedBlockRoot) {
+
+    SafeFuture<Optional<Bytes32>> bestBlockHashPromise =
+        recentChainData
+            .retrieveBlockByRoot(bestBlockRoot)
+            .thenApply(ExecutionEngineService::getPowBlockHash);
+
+    SafeFuture<Bytes32> finalizedBlockHashPromise =
+        maybeFinalizedBlockRoot
+            .map(recentChainData::retrieveBlockByRoot)
+            .orElseGet(() -> SafeFuture.completedFuture(Optional.empty()))
+            .thenApply(ExecutionEngineService::getPowBlockHash)
+            .thenApply(maybeHash -> maybeHash.orElse(Bytes32.ZERO));
```

Comment:
Why can't we do a Noop in the case when finalized block is not yet appeared in the system? In the vein as it is done of the head of the chain.

Reply:
I though it was explicitly states in your [API draft](https://hackmd.io/@n0ble/consensus_api_design_space#Minimal-set-of-methods) and looks like it makes sense: 
> Note: As it is stated by the EIP-3675 the finalized_block_hash must be stubbed with all zeros before the first execution block is getting finalized.

Reply:
This line looks good to me 👍  I want to believe that my comment was targeting another line :-D